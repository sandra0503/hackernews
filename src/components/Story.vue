<template>
  <div class="story" @click="onClick" :class="{ 'story--active': isActive }">
    <div class="story__score">
      <span
        :class="{ story__vote: true, 'story__vote--upvoted': isUpvoted }"
        @click.stop.prevent="onUpvote"
      />
      <span v-if="score" class="story__score-value">{{ score }}</span>
    </div>
    <div class="story__content">
      <div class="story__title">{{ strippedTitle }}</div>
      <div class="story__author">
        by
        <span v-if="author" class="story__author-name">{{ author }}</span>
      </div>
      <div v-if="strippedText" class="story__text" v-html="strippedText"></div>
      <a v-if="linkUrl" class="story__link" target="__blank" :href="linkUrl">
        {{ linkUrl }}
      </a>
    </div>
  </div>
</template>

<script>
export default {
  name: "Story",
  data() {
    return {
      isActive: false,
      isUpvoted: false
    };
  },
  props: {
    title: {
      type: String
    },
    author: {
      type: String
    },
    score: {
      type: Number
    },
    url: {
      type: String
    },
    text: {
      type: String
    }
  },
  computed: {
    strippedTitle() {
      if (!this.isActive && this.title && this.title.length > 90) {
        return `${this.title.slice(0, 89)}...`;
      }
      return this.title;
    },
    strippedText() {
      if (!this.isActive && this.text && this.text.length > 43) {
        return `${this.text.slice(0, 42).replace(/[\r\n<p></p>]/g, " ")}...`;
      }
      return this.text;
    },
    linkUrl() {
      return this.isActive ? this.url : null;
    }
  },
  methods: {
    onClick() {
      this.isActive = !this.isActive;
    },
    onUpvote() {
      this.isUpvoted = true;
    }
  }
};
</script>

<style lang="scss" scoped="true">
.story {
  display: flex;
  padding: 10px;
  overflow: hidden;
  transition: background 0.3s;
  border: 1px solid rgba(232, 236, 241, 1);
  border-radius: 3px;
  cursor: pointer;

  &:hover {
    background-color: rgba(232, 236, 241, 0.5);
  }

  &.story--active {
    grid-area: span 2 / span 1;
    background-color: rgba(232, 236, 241, 0.5);
  }
}

.story__score {
  width: 26px;
  display: flex;
  flex-direction: column;
  text-align: center;
  margin-right: 10px;
  font-size: 12px;
}

.story__vote {
  padding: 5px;
  margin-top: -15px;
  align-self: center;

  &:before {
    content: "";
    display: inline-flex;
    width: 0;
    height: 0;
    border-left: 8px solid transparent;
    border-right: 8px solid transparent;
    border-bottom: 16px solid #009688;
    transition: border 0.2s;
  }

  &:hover {
    &:before {
      border-bottom-color: #00ccb8;
    }
  }

  &.story__vote--upvoted {
    &:before {
      border-bottom-color: #00ccb8;
    }
  }
}

.story__content {
  font-size: 14px;
}

.story__title {
  font-size: 15px;
  font-weight: 500;
}

.story__link {
  word-break: break-all;
  text-decoration: none;
  color: #009688;

  &:hover {
    color: #00ccb8;
  }
}

.story__author {
  color: #808080;
  margin-bottom: 5px;
}

.story__text {
  margin-bottom: 10px;
}

@media only screen and (min-width: 768px) {
  .story {
    &.story--active {
      grid-area: span 2 / span 2;
    }
  }
}
</style>
