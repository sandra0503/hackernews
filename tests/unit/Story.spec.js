import { shallowMount } from "@vue/test-utils";
import Story from "@/components/Story";
import storyMock from "@/__mocks__/story";

describe("Story component", () => {
  let props;

  const build = () => {
    const wrapper = shallowMount(Story, {
      propsData: props
    });
    return {
      wrapper,
      story: () => wrapper.find(".story"),
      title: () => wrapper.find(".story__title"),
      author: () => wrapper.find(".story__author"),
      score: () => wrapper.find(".story__score"),
      link: () => wrapper.find(".story__link"),
      text: () => wrapper.find(".story__text")
    };
  };

  beforeEach(() => {
    const { title, author, score, url, text } = storyMock[0];
    props = {
      title,
      author,
      score,
      url,
      text
    };
  });

  it("renders the component", () => {
    const { wrapper } = build();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders main content elements", () => {
    const { title, author, score, link, text } = build();
    expect(title().exists()).toBe(true);
    expect(author().exists()).toBe(true);
    expect(score().exists()).toBe(true);
    expect(link().exists()).toBe(false);
    expect(text().exists()).toBe(true);
  });

  it("toggles its active state on click", () => {
    const { story } = build();

    expect(story().classes()).not.toContain("story--active");

    story().trigger("click");
    expect(story().classes()).toContain("story--active");

    story().trigger("click");
    expect(story().classes()).not.toContain("story--active");
  });

  it("renders url and/or complete text only in active state ", () => {
    const { story, link } = build();
    story().trigger("click");

    expect(link().exists()).toBe(true);
    expect(link().attributes().href).toBe(props.url);
  });

  it("handles upvote on click of upvote symbol", () => {
    const { score } = build();
    const vote = score().find(".story__vote");
    expect(vote.exists()).toBe(true);

    vote.trigger("click");
    expect(vote.classes()).toContain("story__vote--upvoted");
  });
});
