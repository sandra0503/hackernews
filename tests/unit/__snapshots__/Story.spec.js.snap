// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story component renders the component 1`] = `
<div class="story">
  <div class="story__score"><span class="story__vote"></span> <span class="story__score-value">104</span></div>
  <div class="story__content">
    <div class="story__title">My YC app: Dropbox - Throw away your USB drive</div>
    <div class="story__author">
      by
      <!---->
    </div>
    <div class="story__text">Lorem ipsum</div>
    <!---->
  </div>
</div>
`;
